wikimedia/wikipedia-ios
# Notifications - pull unread side project notifications and batch delete upon logout

**Phabricator:** 
https://phabricator.wikimedia.org/T287298
https://phabricator.wikimedia.org/T297639

### Notes
This wraps up the most important pieces from https://phabricator.wikimedia.org/T287298, which are to scoop up additional unread notifications from projects that are not app languages, commons or wikidata, as well as batch deleting persisted notification data on logout. The notifications API supports a `notcrosswikisummary=1` parameter which returns a list of projects that contain unread notifications. I am pulling this during the primary app language import and refresh operations, then as a follow-on fetching any unread notifications from any new projects determined from this summary object.

The batch delete work simply has the model controller listen for the logout NSNotification and batch deletes the relevant data.

There's one major TODO in here still. It is to rework how we name projects as a bundled list of translations generated from Translatewiki's [Wikimedia Project Names](https://translatewiki.net/w/i.php?title=Special:Translate&group=ext-wikimediaprojectnames&filter=&optional=0&action=page&language=en) project, to prevent unnecessary translations work. I spun off this 2nd one into its own [task](https://phabricator.wikimedia.org/T297620) since that seems like it's own undertaking.

### Test Steps
1. Log into an account that has unread notifications from multiple language wikis. On your primary language desktop, go to Account Preferences > Notifications and be sure "Cross-wiki notifications" checkmark is ON. 

![Screen Shot 2021-12-13 at 9 40 13 AM](https://user-images.githubusercontent.com/3620196/145842254-1c49fcd5-83e4-4f51-98e8-f05e2b8d400d.png)

2. Fresh install app and proxy to watch traffic. Only set up one primary language in App Settings. Go to notifications center, in traffic you should see notifications fetched from wikidata, commons, and your primary app language. You should **also** now see notifications fetched from any other language projects that have unread notifications, even though you do not have them set up as an app language.
3. Back out of Notifications Center and return. In the refresh calls confirm you see notifications fetched from any other language projects that have unread notifications.

4. Now test batch delete on logout. If you are testing in Simulator, you can open the app container printed out in the console and open the RemoteNotifications.sqlite database in your preferred program. You should see notifications in the database after visiting Notifications Center. Log out in simulator via app settings. Refresh database and notice all notifications have now been deleted. Log back in and revisit notifications center. Refresh database and notice all notifications have been re-imported.
